feat: Add sign message support

Strnadj opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

# Checklist:
<!-- CHECKLIST_TYPE: ALL -->
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
<!-- /CHECKLIST_TYPE -->

# Type of change.
<!-- CHECKLIST_TYPE: ONE -->
- [ ] FIX - a PR of this type patches a bug.
- [x] FEATURE - a PR of this type introduces a new feature.
- [ ] BUILD - a PR of this type introduces build changes.
- [ ] CI - a PR of this type introduces CI changes.
- [ ] DOCS - a PR of this type introduces DOCS improvement.
- [ ] STYLE - a PR of this type introduces style changes.
- [ ] REFACTOR - a PR of this type introduces refactoring.
- [ ] PERFORMANCE - a PR of this type introduces performance changes.
- [ ] TEST - a PR of this type adds more tests.
- [ ] CHORE - a PR introduces other changes than the specified above.
<!-- /CHECKLIST_TYPE -->

kujtimprenkuSQA commented on packages/my-near-wallet/src/lib/my-near-wallet.ts:
```diff
@@ -160,6 +160,32 @@ const MyNearWallet: WalletBehaviourFactory<
+      const encodedUrl = encodeURIComponent(url);
+
+      window.location.replace(
+        `${params.walletUrl}/sign-message?message=${message}&nonce=${nonce}&state=${state}&recipient=${recipient}&callbackUrl=${encodedUrl}`
```
After some testing, I found out that depending on what value for `nonce` we pass to the URL it might break the URL structure.

For example trying to sign a message with a nonce like this:

```ts
const nonce = Buffer.from(Array.from(Array(32).keys()));
```

We see this error on the wallet:
![image](https://github.com/near/wallet-selector/assets/95851345/0587f136-2993-4441-acb9-41d7809fc5af)

But if the `nonce` is created from a 32-character long string then the wallet shows the correct UI:
```ts
const nonce = Buffer.from("abcdefghijklmnopqrstuvwxyz123456");
```
![image](https://github.com/near/wallet-selector/assets/95851345/60cf6f47-943a-4bbf-825c-2c7f68afadf5)

kujtimprenkuSQA reviewed (commented):
Hi, @Strnadj thank you for submitting this PR.

Can you please change the base branch from `main` to `dev`:
https://github.com/near/wallet-selector/blob/main/CONTRIBUTING.md#branching

After testing it out I noticed that the `data` (accountId, publicKey, signature, state) are being returned as query params:

```bash
http://localhost:4200/?accountId=kujtimprenku.near&signature=chzNwZMcPopCrGxHF7j78qhw8KFygXVJS7SYug7lPPrEDkjt%2BNNtFLOxi%2Bq%2BSptOBD0vDsU2QHxIY8XSrebvCg%3D%3D&publicKey=ed25519%3AHiqgEJAB5AWUKTtrowxqBYCfzqpdFbDz2LvQND36bAUY&state=undefined
```

The NEP states that this `data` should be returned in a Fragment URL for Browser Wallets:
https://github.com/near/NEPs/blob/master/neps/nep-0413.md#web-wallets
https://github.com/near/NEPs/blob/master/neps/nep-0413.md#how-to-return-the-signed-message-in-a-safe-way

```bash
# here only `?` is replaced with `#`
http://localhost:4200/#accountId=kujtimprenku.near&signature=chzNwZMcPopCrGxHF7j78qhw8KFygXVJS7SYug7lPPrEDkjt%2BNNtFLOxi%2Bq%2BSptOBD0vDsU2QHxIY8XSrebvCg%3D%3D&publicKey=ed25519%3AHiqgEJAB5AWUKTtrowxqBYCfzqpdFbDz2LvQND36bAUY&state=undefined
```

Strnadj commented on packages/my-near-wallet/src/lib/my-near-wallet.ts:
```diff
@@ -160,6 +160,32 @@ const MyNearWallet: WalletBehaviourFactory<
+
+      if (!account) {
+        throw new Error("Wallet not signed in");
+      }
```
✅

Strnadj commented on packages/my-near-wallet/src/lib/my-near-wallet.spec.ts:
```diff
@@ -132,3 +132,29 @@ describe("buildImportAccountsUrl", () => {
+
+    const replace = window.location.replace;
+
+    await wallet.signIn({ contractId: "test.testnet" });
```
✅

kujtimprenkuSQA commented:
@Strnadj thank you for addressing the comments, all that's left is to return the signed-message (accountId, publicKey, signature, state) in a URL Fragment to align the with the NEP.

I assume this will need an update in the MyNearWallet.

Strnadj commented:
> @Strnadj thank you for addressing the comments, all that's left is to return the signed-message (accountId, publicKey, signature, state) in a URL Fragment to align the with the NEP.
> 
> I assume this will need an update in the MyNearWallet.

Yes, I am working on that.

Strnadj commented:
@kujtimprenkuSQA I think this is ready for review ... together with https://github.com/mynearwallet/my-near-wallet/pull/164/files

kujtimprenkuSQA commented:
> @kujtimprenkuSQA I think this is ready for review ... together with https://github.com/mynearwallet/my-near-wallet/pull/164/files

Thanks again for making the improvements/changes I checked the PR in MyNearWallet everything looks good to me.
We will publish a new version of the wallet-selector as soon as this is live in production in MNW.

Strnadj commented:
@kujtimprenkuSQA the PR with MNW is merged.
